### PR TITLE
Cleanup arg{min, max}

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Why do we need this workaround? `PythonArgParser` handles these two cases well.

The discussion started at https://github.com/pytorch/pytorch/pull/6201#issuecomment-378724406. The conclusion at that time by @goldsborough was:

> Because we wanted to allow `dim=None` in Python and route to a different function. Essentially the problem was wanting to wrap the C++ function in Python. AFAIK there is no way of translating `dim=None` behavior into C++? So Richard and I came up with this strategy

Maybe at that time `PythonArgParser` was not powerful enough to handle the routing of two function with same name but different C++ signature.

Will keep an eye on the CI.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Too many `E: Could not get lock /var/lib/apt/lists/lock - open (11: Resource temporarily unavailable)` on CI, reopening...

**Comment by zasdfgbnm** [comment]
Too many `E: Could not get lock /var/lib/apt/lists/lock - open (11: Resource temporarily unavailable)` failures on circle CI..

**Comment by zasdfgbnm** [comment]
Yes, all tests pass. We no longer need that workaround. Ready for review now.

**Comment by colesbury** [comment]
The comment says "we wanted to allow `dim=None` in Python". This no longer works with the PR. (It appears we are missing a test case for this)

```python
>>> torch.randn(5, 5).argmax(dim=None)
TypeError: argmax(): argument 'dim' must be int, not NoneType
```

**Comment by zasdfgbnm** [comment]
@colesbury Please see the new PR https://github.com/pytorch/pytorch/pull/17306 which support `dim=None`. But why are we interested in writing `x.argmax(dim=None)` instead of `x.argmax()`? IMHO, the only reason that we support `dim=None` was to find a way for the user to route to the flatten version of argmax, which historically I guess was done by setting the default value `dim=None` in python side and check if it was `None` then route to the flatten version. But now the python arg parser handles this route correctly, so we no longer need to support `x.argmax(dim=None)`. So personally I would close #17306 in favor of this.

**Comment by ezyang** [comment]
While I'd agree that there's probably no one who would actually want to do things this way, contracts are contracts, and there's no reason to gratuitously break people in this case. In the interest of not having an intermediate broken state; may I suggest squashing these two PRs?

**Comment by zasdfgbnm** [comment]
@ezyang Yes, I will squash this with https://github.com/pytorch/pytorch/pull/17306. The only thing is, adding support for `dim=None` breaks the shape analysis, I'm not sure how to fix that yet, just give me some time to look deep into it.

**Comment by zasdfgbnm** [comment]
@ezyang I have squashed the two PR and fixed shape analysis. This PR is now ready for review.

File: aten/src/ATen/native/TensorCompare.cpp [hunk 1]
```diff
@@ -217,32 +217,18 @@ std::tuple<Tensor &,Tensor &> min_out(Tensor& min, Tensor& min_indices,
...
 
-Tensor argmax(const Tensor& self, int64_t dim, bool keepdim) {
-  return std::get<1>(self.max(dim, keepdim));
-}
-
-Tensor argmax(const Tensor& self) {
+Tensor argmax(const Tensor& self, c10::optional<int64_t> dim, bool keepdim) {
+  if (dim)
+    return std::get<1>(self.max(dim.value(), keepdim));
+  AT_CHECK(!keepdim, "keepdim is meaningful only when dim != None")
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/TensorCompare.cpp]: This behavior doesn't match documentation: "Ignored if ``dim=None``"

File: aten/src/ATen/native/TensorCompare.cpp [hunk 2]
```diff
@@ -217,32 +217,18 @@ std::tuple<Tensor &,Tensor &> min_out(Tensor& min, Tensor& min_indices,
...
 
-Tensor argmin(const Tensor& self, int64_t dim, bool keepdim) {
-  return std::get<1>(self.min(dim, keepdim));
-}
-
-Tensor argmin(const Tensor& self) {
+Tensor argmin(const Tensor& self, c10::optional<int64_t> dim, bool keepdim) {
+  if (dim)
+    return std::get<1>(self.min(dim.value(), keepdim));
+  AT_CHECK(!keepdim, "keepdim is meaningful only when dim != None")
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/TensorCompare.cpp]: This behavior doesn't match documentation: "Ignored if ``dim=None``"

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]

**Comment by zasdfgbnm** [comment]
@VitalyFedyunin fixed

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 3]
```diff
@@ -970,6 +968,33 @@ class ShapePropagator {
...
+            "aten::argmin(Tensor self, int? dim, bool keepdim) -> Tensor",
+        },
+        [](Node* node) -> type_vec_t {
+          if (auto type =
+                  node->input(0)->type()->cast<DimensionedTensorType>()) {
+            if (node->input(1)->type()->kind() == c10::TypeKind::NoneType) {
+              return {type->withDim(0)};
+            } else {
+              return multidim_reduce_with_postprocess(
+                  node, /*num_reduce_dim=*/1, /*integer_upcast=*/false);
```
Comment (VitalyFedyunin) [inline torch/csrc/jit/passes/shape_analysis.cpp]: nit: 
  Argument name in comment (num_reduce_dim) does not match parameter name (num_reduced_dim).
  Argument name in comment (integer_upcast) does not match parameter name (upcast_integer).

Comment (zasdfgbnm) [reply torch/csrc/jit/passes/shape_analysis.cpp]: @VitalyFedyunin fixed

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Thanks for cleaning up!

**Comment by houseroad** [comment]
@VitalyFedyunin @zasdfgbnm the onnx CI failed, and it's legit...

**Comment by zasdfgbnm** [comment]
@houseroad Maybe PyTorch is missing test for ONNX exporter of argmax and argmin?

**Comment by zasdfgbnm** [comment]
@houseroad @VitalyFedyunin I think the problem is at https://github.com/pytorch/pytorch/blob/master/torch/onnx/symbolic.py#L1692

```python
@parse_args('v', 'i', 'i')
def _argmax(g, input, dim, keepdim):
    return g.op('ArgMax', input, axis_i=dim, keepdims_i=keepdim)


@parse_args('v', 'i', 'i')
def _argmin(g, input, dim, keepdim):
    return g.op('ArgMin', input, axis_i=dim, keepdims_i=keepdim)
```

Is it OK to allow me one day (tomorrow, March 21) to write a fix? If I have some trouble and unable to get it fixed tomorrow, then we might just revert this diff.

**Comment by houseroad** [comment]
@zasdfgbnm I will take care of the onnx part this time, please please do check the failing CI next time!!!

the fix is here: https://github.com/pytorch/pytorch/pull/18261

**Comment by zasdfgbnm** [comment]
@houseroad Sorry for not checking the onnx test. I though the `caffe2_onnx_py2_gcc5_ubuntu16_04` I only need to care about this if I modified caffe2...

**Pull request closed without merging**